bcdaworker/queueing/manager/alr.go

```diff
@@ -19,64 +27,199 @@ type alrQueue struct {
 	alrWorker worker.AlrWorker
 }
 
+// alrJobTrackerStruct is used to track the status of the job
+type alrJobTrackerStuct struct {
+	tracker map[uint]alrJobAttempt // unint here is the big job
+	sync.Mutex
+}
+
+type alrJobAttempt struct {
+	status         models.JobStatus
+	attemptTracker map[uint]uint // 1st uint is the small job
+}
+
+// Instance at the global private level
+var alrJobTracker = alrJobTrackerStuct{
+	tracker: make(map[uint]alrJobAttempt),
+}
+
+/******************************************************************************
+	Functions
+******************************************************************************/
+
+// checkIFCanncelled was originally a closure to check if the job was cancelled,
+// but it has been turned into a func for ALR for clarity
+func checkIfCancelled(ctx context.Context, q *masterQueue,
+	cancel context.CancelFunc, jobArgs models.JobAlrEnqueueArgs) {
+	for {
+		select {
+		case <-time.After(15 * time.Second):
+			jobStatus, err := q.repository.GetJobByID(ctx, jobArgs.ID)
+
+			if err != nil {
+				q.alrLog.Warnf("Could not find job %d status: %s", jobArgs.ID, err)
+			}
+
+			if jobStatus.Status == models.JobStatusCancelled {
+				cancel()
+				return
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// deleteAndRetry is a helper function that will delete previous attempt at
+// creating ndjson if the worker hit an error along the way. This is to avoid
+// serving incomplete or redundant data to user.
+func deleteAndRetry(l *logrus.Logger, dir string, id uint, file string) {
+	err := os.Remove(fmt.Sprintf("%s/%d/%s.ndjson", dir, id, file))
+	if err != nil {
+		l.Fatal("Could not remove alr ndjson.")
+	}
+}
+
+/******************************************************************************
+	Methods
+******************************************************************************/
+
 // startALRJob is the Job that the worker will run from the pool. This function
 // has been written here (alr.go) to separate from beneficiary FHIR workflow.
 // This job is handled by the same worker pool that works on beneficiary.
 func (q *masterQueue) startAlrJob(job *que.Job) error {
-	// Creating Context for possible cancellation
+
+	// Creating Context for possible cancellation; used by checkIfCancelled fn
 	ctx := context.Background()
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	// Unmarchall JSON that contains the job details
-	// JSON selected for ALR for continuity... may change in the future
+	// Unmarshall JSON that contains the job details
 	var jobArgs models.JobAlrEnqueueArgs
 	err := json.Unmarshal(job.Args, &jobArgs)
+	// If we cannot unmarshall this, it would be very problematic.
+	if err != nil {
+		// TODO: perhaps just fail the job?
+		q.alrLog.Fatalf("Failed to unmarhall job.Args '%s' %s.",
+			job.Args, err)
+	}
+
+	// Check if this is already a failed job, and don't continue if it is
+	alrJobs, err := q.repository.GetJobByID(ctx, jobArgs.ID)
```

Comment:
Since we commit the transaction that contains the new model.Job, we may encounter a situation where we have the queue jobs but not the parent job yet.

We handle it like this in the que.go manager now. https://github.com/CMSgov/bcda-app/blob/master/bcdaworker/queueing/manager/que.go#L134 can we follow a similar approach?